Add Codeflow PR User guide to UB documentation (#15643)
dotnet/arcade: Documentation/UnifiedBuild/CodeflowPrUserGuide.md

## Codeflow PR Metadata

The Codeflow mechanism relies on metadata files that can be found in each dotnet repository that is managed by the PCS.
- **`source-manifest.json`**
  - Records the origin of source code changes that are synchronized into the VMR.
- **`Version.Details.xml`**
  - Contains versioning information for dependencies, ensuring consistent package updates.

## What to Do in Case of Conflicts

If a Codeflow PR encounters merge conflicts, follow these steps:
[TBD]

## FAQ

- **Where do packages get built?**
  - Packages are built in Unified Build pipelines. See [UB Pipelines](https://dev.azure.com/dnceng/internal/_build?definitionId=1330).
- **How do I find the new codeflow subscriptions?**
  - A full list of subscriptions can be found at on the [maestro.dot.net](https://maestro.dot.net/subscriptions) webpage. Subscriptions can be managed with DARC commands in the same way they always were.
- **What should I do in case of conflicts?**

>   - Follow the steps in [What to Do in Case of Conflicts](#what-to-do-in-case-of-conflicts).

Reviewer: @adamzip this one was missed maybe

>   - Follow the steps in [What to Do in Case of Conflicts](#what-to-do-in-case-of-conflicts).

Reviewer: @adamzip this one is still not resolved. Please copy the whole FAQ (or most of it if some questions don't make sense but I'd say most do)

> # Codeflow PR User Guide
> 
> # Codeflow PRs

Reviewer: ```suggestion
# Codeflow PRs
```

Codeflow PRs are automated pull requests created by [Maestro](https://maestro.dot.net), and are generated by source-enabled subscriptions.  
These subscriptions can only be created between the [VMR (Virtual Monolithic Repository)](https://github.com/dotnet/dotnet/) and product repositories that have their [sources in the VMR](https://github.com/dotnet/dotnet/tree/main/src).  

As the name suggests, source-enabled subscriptions not only update dependencies but also synchronize source code changes **to and from** the VMR.
They ensure that repository sources remain in sync with the corresponding `src/` directory in the VMR.
Source code changes can originate in either the repository or the VMR and are synchronized to the other side via source-enabled subscriptions.

> > [!IMPORTANT]  
> > Opening PRs against the VMR is **not yet permitted** and will be allowed in the near future. For now, source code changes should only be made in the repositories, barring exceptional cases.  
> 
> For more details on codeflow and the VMR, see [VMR Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).  

Reviewer: ```suggestion
> [!IMPORTANT]
> Opening PRs against the VMR is **not yet permitted** and will be allowed in the near future. For now, source code changes should only be made in the repositories, barring exceptional cases.

For more details on codeflow and the VMR, see [VMR Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).
```

> ## Codeflow PR Metadata

Reviewer: ```suggestion
## Terminology

- **Product repository** / **VMR repository** - A repository that is required to build the .NET SDK, and which is synchronized [into the VMR](https://github.com/dotnet/dotnet/tree/main/src).
- **Dependency flow** / **Binary flow** - Old type of Maestro subscriptions that are only flowing dependency updates, e.g. [this one](https://github.com/dotnet/sdk/pull/47085).
- **Code flow** / **Source-enabled dependency flow** - New type of Maestro subscriptions that are, together with dependency updates, also flowing sources to/from the VMR.
- **Backflow** - A PR that flows changes from the VMR back to the product repository, carrying source updates as well as dependency updates (packages built in the VMR).
- **Forward flow** - A PR that flows changes from the product repository to the VMR, carrying repository source updates.
- **Flat flow** - A new structure of subscriptions between product repositories and the VMR.

## Codeflow PR Metadata
```

Author: Should VMR repository really be another way to call product repos? That sounds confusing

Reviewer: I think it's already used around in other docs and we use it quite commonly too

# Codeflow PR User Guide

## What is Codeflow?

> Codeflow PRs are created by source-enabled subscriptions, which are a new type of [Maestro dependency subscriptions]. 
> 
> 

Reviewer: ```suggestion
```
This was left behind?

> Codeflow PRs are automated pull requests created by [Maestro](https://maestro.dot.net). They are a new type of automated dependency PRs that are created by source-enabled subscriptions. These subscriptions can be only created between a the [VMR (Virtual Monolithic Repository)](https://github.com/dotnet/dotnet/) and product repositories that have their [sources in the VMR](https://github.com/dotnet/dotnet/tree/main/src)

Reviewer: ```suggestion
Codeflow PRs are automated pull requests created by [Maestro](https://maestro.dot.net).
They are a new type of automated dependency PRs that are created by source-enabled subscriptions.
These subscriptions can be only created between a the [VMR (Virtual Monolithic Repository)](https://github.com/dotnet/dotnet/) and product repositories that have their [sources in the VMR](https://github.com/dotnet/dotnet/tree/main/src)
```

For markdown, we usually put one sentence per line as it does not have effect on the final view.

> Codeflow PRs have two roles:
> - Managing the dependencies that exist among dotnet repositories,
> - Synchronizing all of the source code from dotnet repositories into the [VMR](https://github.com/dotnet/dotnet#readme). 
> 
> Previously, automated PRs were used to update dependencies among dotnet repositoires, but source code was not synchronized. The introduction of the VMR allows us to build the entirety of the dotnet sdk rapidly from a single source.
> 
> Another consequence of Codeflow PRs is that the intricate dependency graph among dotnet repositories is replaced by a flat graph of dependencies (also called Flat Flow), where each product subscribes only to updates coming from the VMR.
> 
> For additional information about codeflow and the VMR, refer to [VMR Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).

Reviewer: I don't think it's true - it synchronizes the code between them, not just into the VMR.

```suggestion
As the name hints, the source-enabled subscriptions do not only contain dependency updates, but also bring over the source code changes.
They synchronize the sources of repositories with the contents of the matching `src/` directory in the VMR.
This way, a change can be made either in the repository or in the VMR and the next codeflow PR will bring this change over to the other side.

> [!IMPORTANT]
> Please note that opening PRs against the VMR is not yet permitted and will be allowed in the near future. So for now, the changes are mostly made in the repositories. From there they are forward-flown to the VMR, where the code is built, and the newly built dependencies are flown back to the repositories (via source-enabled backflow subscriptions).

For additional information about codeflow and the VMR, refer to [VMR Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).
```

Author: I ran it through chatgpt for some small rewordings and also added a clarification that source-enabled subscriptions are only to and from the VMR, so that people don't get the idea that source code can flow between product repos

## Codeflow PR Metadata

> The Codeflow mechanism relies on metadata files that can be found in each dotnet repository that is managed by the PCS. These files should not be modified by humans!
> - **`source-manifest.json`**
>   - A file contained in the [VMR](https://github.com/dotnet/dotnet) that records the latest source-code synchronization between every repository and the VMR.
> - **`Version.Details.xml`**
>   - This file already existed - it contains versioning information for dependencies between repositories. A new <Source> tag is added to track the latest codeflow from the VMR.

Reviewer: ```suggestion
The code flow mechanism relies on metadata files similarly to the already existing Maestro dependency flow:
- **`eng/Version.Details.xml`**
  - A file you already know from the current dependency flow. It contains version information for dependencies between repositories. Newly, a `<Source>` tag is added to track the latest codeflow from the VMR.
- **`src/source-manifest.json`**
  - A file contained in the [VMR](https://github.com/dotnet/dotnet/blob/main/src/source-manifest.json) that contains the last version of every repository synchronized into the VMR.
```

# Codeflow PR User Guide

# Codeflow PRs

Codeflow PRs are automated pull requests created by [Maestro](https://maestro.dot.net), and are generated by source-enabled subscriptions.  
These subscriptions can only be created between the [VMR (Virtual Monolithic Repository)](https://github.com/dotnet/dotnet/) and product repositories that have their [sources in the VMR](https://github.com/dotnet/dotnet/tree/main/src).  

> As the name suggests, source-enabled subscriptions not only update dependencies but also synchronize source code changes **to and from** the VMR.  
> They ensure that repository sources remain in sync with the corresponding `src/` directory in the VMR.  
> Source code changes can originate in either the repository or the VMR and are synchronized to the other side via source-enabled subscriptions.  

Reviewer: ```suggestion
As the name suggests, source-enabled subscriptions not only update dependencies but also synchronize source code changes **to and from** the VMR.
They ensure that repository sources remain in sync with the corresponding `src/` directory in the VMR.
Source code changes can originate in either the repository or the VMR and are synchronized to the other side via source-enabled subscriptions.
```

> [!IMPORTANT]  
> Opening PRs against the VMR is **not yet permitted** and will be allowed in the near future. For now, source code changes should only be made in the repositories, barring exceptional cases.  

> For more details on codeflow and the VMR, see [VMR Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).  

Reviewer: ```suggestion
> [!IMPORTANT]
> Opening PRs against the VMR is **not yet permitted** and will be allowed in the near future. For now, source code changes should only be made in the repositories, barring exceptional cases.

For more details on codeflow and the VMR, see [VMR Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).
```
